pairtools/pairtools_dedup.py

```diff
@@ -274,14 +450,209 @@ def fetchadd(key, mydict):
+    else:
+        p = np.inf
+
+    unmapped_id = (df["chrom1"] == unmapped_chrom) | (df["chrom2"] == unmapped_chrom)
+    unmapped = df[unmapped_id]
+
+    df = df[~unmapped_id]
+    N = df.shape[0]
+    if N > 0:
+        if backend == "sklearn":
+            a = neighbors.radius_neighbors_graph(
+                df[["pos1", "pos2"]],
+                radius=r,
+                p=p,
+                n_jobs=n_proc,
+            )
+            a0, a1 = a.nonzero()
+        elif backend == "scipy":
+            z = scipy.spatial.cKDTree(df[["pos1", "pos2"]])
+            a = z.query_pairs(r=r, p=p, output_type="ndarray")
+            a0 = a[:, 0]
+            a1 = a[:, 1]
+        need_to_match = np.array(
+            [
+                ("chrom1", "chrom1"),
+                ("chrom2", "chrom2"),
+                ("strand1", "strand1"),
+                ("strand2", "strand2"),
+            ]
+            + extra_col_pairs
+        )
+        nonpos_matches = np.all(
+            [
+                df.iloc[a0, df.columns.get_loc(lc)].values
+                == df.iloc[a1, df.columns.get_loc(rc)].values
+                for (lc, rc) in need_to_match
+            ],
+            axis=0,
+        )
+        a0 = a0[nonpos_matches]
+        a1 = a1[nonpos_matches]
+        a_mat = coo_matrix((np.ones_like(a0), (a0, a1)), shape=(N, N))
+
+        df["clusterid"] = connected_components(a_mat, directed=False)[1]
+
+    else:
+        df["clusterid"] = np.nan
+    dups = df["clusterid"].duplicated()
+    df["duplicate"] = False
+    if keep_parent_read_id:
+        df["parent_readID"] = df["clusterid"].map(
+            df[~dups].set_index("clusterid")["readID"]
+        )
+        unmapped["parent_readID"] = ""
+    unmapped["duplicate"] = False
+
+    df.iloc[dups, df.columns.get_loc("duplicate")] = True
+    return pd.concat([unmapped, df.drop(["clusterid"], axis=1)]).reset_index(drop=True)
+
+
+def _dedup_by_chunk(
+    in_stream,
+    colnames,
+    method,
+    chunksize,
+    mark_dups,
+    max_mismatch,
+    extra_col_pairs,
+    save_parent_id,
+    comment_char,
+    backend,
+    n_proc,
+):
+    dfs = pd.read_table(
+        in_stream, comment=comment_char, names=colnames, chunksize=chunksize
+    )
+
+    old_nodups = pd.DataFrame([])
+    old_i = 0
+    for df in dfs:
+        marked = dedup_chunk(
+            pd.concat([old_nodups, df], axis=0, ignore_index=True).reset_index(
+                drop=True
+            ),
+            r=max_mismatch,
+            method=method,
+            keep_parent_read_id=save_parent_id,
+            extra_col_pairs=extra_col_pairs,
+            backend=backend,
+            n_proc=n_proc,
+        )
+        marked = marked.iloc[old_i:, :].reset_index(drop=True)
+        if mark_dups:
+            marked.iloc[marked["duplicate"], marked.columns.get_loc("pair_type")] = "DD"
+        nodups = marked[~marked["duplicate"]]
+
+        nodups = nodups[colnames]
+        i = max(nodups.shape[0] // 100, 100)
```

Comment:
Several suggestions: 
1. Upgrade 1% and 100 of nodups of the previous chunk into a parameter. 
2. If the chunk is too large, then it does not make sense to take a huge number of pairs. On the other hand, if the datasets is too low complexity, 100 pairs might be not enough. 
Is it possible to make some "iterative prepending" instead? Like, take only the bottom values that have 2-4 different parent readIDs. This will guarantee we don't miss anything and do not make extra work.

Reply:
I don't think 100 is ever not enough realistically, since it only takes already deduped pairs, so they are suppoed to all be distinct. What I was thinking is we probably need even fewer...

Reply:
Just group of last ~4 distinct pairs? The number 4 is the number of possible strand orientations combinations here, I'm not sure it's needed.

Reply:
Interesting. Maybe that would be enough... I am struggling to imagine a case where it wouldn't be. But so few sounds scary!

Reply:
So, we decided to make it a parameter defaulting to 100.